core/shared/src/main/scala/fs2/internal/FreeC.scala

```diff
@@ -12,138 +12,223 @@ import scala.util.control.NonFatal
+      case Result.Fail(err) =>
+        try { f(err) } catch { case NonFatal(err2) => Result.Fail(CompositeFailure(err, err2)) }
+      case _ => covary[R2]
+    }
+
+  }
+
+  object Result {
+
+    val unit: Result[Unit] = pure(())
+
+    def pure[A](a: A): Result[A] = Result.Pure(a)
+
+    def raiseError[A](rsn: Throwable): Result[A] = Result.Fail(rsn)
+
+    def interrupted[A](scopeId: Token, failure: Option[Throwable]): Result[A] =
+      Result.Interrupted(scopeId, failure)
+
+    def fromEither[F[_], R](either: Either[Throwable, R]): Result[R] =
+      either.fold(Result.Fail(_), Result.Pure(_))
+
+    def unapply[F[_], R](freeC: FreeC[F, R]): Option[Result[R]] = freeC match {
+      case r @ Result.Pure(_)           => Some(r: Result[R])
+      case r @ Result.Fail(_)           => Some(r: Result[R])
+      case r @ Result.Interrupted(_, _) => Some(r: Result[R])
+      case _                            => None
+    }
+
+    final case class Pure[F[_], R](r: R) extends FreeC[F, R] with Result[R] {
+      override def translate[G[_]](f: F ~> G): FreeC[G, R] =
+        this.asInstanceOf[FreeC[G, R]]
+      override def toString: String = s"FreeC.Pure($r)"
+    }
+
+    final case class Fail[F[_], R](error: Throwable) extends FreeC[F, R] with Result[R] {
+      override def translate[G[_]](f: F ~> G): FreeC[G, R] =
+        this.asInstanceOf[FreeC[G, R]]
+      override def toString: String = s"FreeC.Fail($error)"
+    }
+
+    /*
+     * Interruption of the stream is tightly coupled between FreeC, Algebra and CompileScope
+     * Reason for this is unlike interruption of `F` type (i.e. IO) we need to find
+     * recovery point where stream evaluation has to continue in Stream algebra
+     *
+     * As such the `Interrupted` is necessary glue between FreeC/Algebra that allows pass-along
+     * information for Algebra and scope to correctly compute recovery point after interruption was signalled via `CompilerScope`.
+     *
+     * Parameter `scope` indicates scope of the computation where interruption actually happened.
+     * This is used to precisely find most relevant interruption scope where interruption shall be resumed
+     * for normal continuation.
+     * Essentially interpreter uses this to find any parents of this scope that has to be interrupted, and guards the
+     * interruption so it won't propagate to scope that shall not be anymore interrupted.
+     *
+     * Parameter `finalizerErrors` allows to collect errors that may accumulate during interruption propagation.
+     * When interruption is executed, all finalizers are executed, until interruption resumes. During this time,
+     * finalizers may accumulate errors. These errors are then stored here, and are raised when interruption resumes.
+     * Reason for this delayed reporting is to allow proper recovery with `handleErrorWith` handlers.
+     *
+     */
+    final case class Interrupted[F[_], R](scope: Token, finalizerErrors: Option[Throwable])
```

Comment:
Maybe this is silly, but I would feel better about `Interrupted` class if the names of `scope` and `finalizerErrors` were more general. For example:

```scala
final case class Interrupted[F[_], R](token: Token, deferredError: Option[Throwable])
```

Or maybe even making the token parameterized:

```scala
final case class Interrupted[F[_], T, R](tag: T, deferredError: Option[Throwable])
```

Reply:
Thanks for the docs by the way -- I'd prefer if we found a place for those docs in Algebra or Stream though. Conceptually, I'd like to keep `FreeC` as a standalone thing with no references to Streams.

Reply:
ok I like idea with type parameters. Not sure how that will play with current pattern match, in worst case I can have unapply variant on Algebra
 

```scala
final case class Interrupted[F[_], A, R](interruptContext: A)
```
and then I may move the explanation/doc to Algebra interpreter?

Reply:
That sounds good